phantazm-neuron-minestom/src/main/java/com/github/phantazmnetwork/neuron/bindings/minestom/entity/NeuralEntity.java

```diff
@@ -0,0 +1,230 @@
+ */
+public abstract class NeuralEntity extends LivingEntity implements Agent {
+    private final MinestomDescriptor descriptor;
+    private final ContextProvider provider;
+
+    private Navigator navigator;
+    private Controller controller;
+    private Explorer explorer;
+
+    private int removalAnimationDelay;
+
+    /**
+     * Creates a new NeuralEntity.
+     * @param descriptor the descriptor used to define this agent's characteristics
+     * @param uuid this entity's unique ID
+     * @param provider the {@link ContextProvider} used to provide {@link PathContext} instances for this entity
+     */
+    public NeuralEntity(@NotNull MinestomDescriptor descriptor, @NotNull UUID uuid, @NotNull ContextProvider provider) {
+        super(descriptor.getEntityType(), uuid);
+        this.descriptor = descriptor;
+        this.provider = Objects.requireNonNull(provider, "provider");
+        this.removalAnimationDelay = 1000;
+    }
+
+    @Override
+    public void preTick(long time) {
+        navigator.tick(time);
+    }
+
+    @Override
+    public boolean canPathfind() {
+        Instance instance = this.instance;
+        Chunk currentChunk = this.currentChunk;
+        return instance != null && currentChunk != null && !isDead() && instance.getWorldBorder().isInside(this)
+                && !instance.isInVoid(getPosition()) && currentChunk.isLoaded() && vehicle == null;
+    }
+
+    @Override
+    public @NotNull Vec3I getStartPosition() {
+        return VecUtils.toBlockInt(getPosition());
+    }
+
+    @Override
+    public @NotNull Descriptor getDescriptor() {
+        return descriptor;
+    }
+
+    @Override
+    public @NotNull Explorer getExplorer() {
+        requireInstance();
+        return explorer;
+    }
+
+    @Override
+    public @NotNull Controller getController() {
+        requireInstance();
+        return controller;
+    }
+
+    /**
+     * Returns the {@link Navigator} managing this entity. Will throw an exception if this entity does not have an
+     * instance defined yet.
+     * @return the Navigator used for pathfinding
+     */
+    public @NotNull Navigator getNavigator() {
+        requireInstance();
+        return navigator;
+    }
+
+    /**
+     * Sets the given entity as a navigation target for this entity. If null, pathfinding will be cancelled. If
+     * non-null, this entity will follow the given entity according to its navigation capabilities.
+     * @param target the target entity to follow
+     */
+    public void setTarget(@Nullable Entity target) {
+        if(target == null) {
+            navigator.setDestination(null);
+        }
+        else {
+            navigator.setDestination(() -> VecUtils.toBlockInt(target.getPosition()));
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> setInstance(@NotNull Instance instance, @NotNull Pos spawnPosition) {
+        return super.setInstance(instance, spawnPosition).whenComplete((ignored, ex) -> {
```

Comment:
if you ignore ex (also maybe you shouldn't) then you could use thenRun

Reply:
thenRun shouldn't execute if super.setInstance throws, which is probably what i want tbh